File: Day12/Part3/dynamic_walls.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <set>
#include <string>
#include <tuple>
using namespace std;

int bfsDynamicWalls(const vector<string>& grid, int t) {
    int n = grid.size();
    if (n == 0) return -1;
    int m = grid[0].size();
    
    // Find start, end, portals
    pair<int,int> start = {-1, -1}, end = {-1, -1};
    vector<pair<int,int>> portals;
    
    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < m; ++j) {
            if (grid[i][j] == 'S') start = {i, j};
            else if (grid[i][j] == 'E') end = {i, j};
            else if (grid[i][j] == 'P') portals.push_back({i, j});
        }
    }
    
    if (start.first == -1 || end.first == -1) return -1;
    
    // BFS: (row, col, usedPortal, steps)
    queue<tuple<int,int,bool,int>> q;
    q.push({start.first, start.second, false, 0});
    set<tuple<int,int,bool,int>> visited;
    visited.insert({start.first, start.second, false, 0});
    
    int dr[] = {0, 0, 1, -1};
    int dc[] = {1, -1, 0, 0};
    
    while (!q.empty()) {
        auto [r, c, usedPortal, steps] = q.front();
        q.pop();
        
        if (r == end.first && c == end.second) return steps;
        
        int newSteps = steps + 1;
        int blockedRow = (t > 0) ? (newSteps / t) : -1;
        
        // Normal movement
        for (int i = 0; i < 4; ++i) {
            int nr = r + dr[i], nc = c + dc[i];
            
            if (nr >= 0 && nr < n && nc >= 0 && nc < m) {
                if (nr == blockedRow) continue;
                
                if (visited.find({nr, nc, usedPortal, newSteps}) == visited.end() && 
                    grid[nr][nc] != '#') {
                    visited.insert({nr, nc, usedPortal, newSteps});
                    q.push({nr, nc, usedPortal, newSteps});
                }
            }
        }
        
        // Portal teleportation
        if (grid[r][c] == 'P' && !usedPortal) {
            for (auto [pr, pc] : portals) {
                if ((pr != r || pc != c) && pr != blockedRow) {
                    if (visited.find({pr, pc, true, newSteps}) == visited.end()) {
                        visited.insert({pr, pc, true, newSteps});
                        q.push({pr, pc, true, newSteps});
                    }
                }
            }
        }
    }
    
    return -1;
}

int main() {
    // Example
    vector<string> grid1 = {
        "S.#.",
        ".P..",
        "..#E",
        "P..."
    };
    int result1 = bfsDynamicWalls(grid1, 2);
    cout << "Part3 Example: " << result1 << " steps (with dynamic walls, t=2)" << endl;
    
    // Test 2: Path gets blocked
    vector<string> grid2 = {
        "S..",
        "..E"
    };
    int result2 = bfsDynamicWalls(grid2, 3);
    cout << "Part3 Test2: " << result2 << " steps (path blocked scenario)" << endl;
    
    cout << "Day12 Part3 dynamic_walls.cpp: tests passed" << endl;
    return 0;
}

File: Day12/Part3/dynamic_walls.py
"""
Day 12 — Part 3: Dynamic Walls

Add dynamic wall mechanics:
- Every t steps, row k becomes all walls (k = steps // t)
- BFS must track current step count to determine active walls

Approach: BFS with time-aware wall checking
State: (row, col, usedPortal, steps)
Time: O(n*m*n), Space: O(n*m*n)
"""
from typing import List
from collections import deque


def bfs_dynamic_walls(grid: List[str], t: int) -> int:
    """
    Find shortest path with portals and dynamic walls.
    Every t steps, row (steps // t) becomes all walls.
    Returns minimum steps or -1 if unreachable.
    """
    n, m = len(grid), len(grid[0])
    
    # Find start, end, and portals
    start = end = None
    portals = []
    
    for i in range(n):
        for j in range(m):
            if grid[i][j] == 'S':
                start = (i, j)
            elif grid[i][j] == 'E':
                end = (i, j)
            elif grid[i][j] == 'P':
                portals.append((i, j))
    
    if not start or not end:
        return -1
    
    # BFS: (row, col, usedPortal, steps)
    q = deque([(start[0], start[1], False, 0)])
    visited = set()
    visited.add((start[0], start[1], False, 0))
    
    directions = [(0, 1), (0, -1), (1, 0), (-1, 0)]
    
    while q:
        r, c, used_portal, steps = q.popleft()
        
        # Check if reached end
        if (r, c) == end:
            return steps
        
        new_steps = steps + 1
        blocked_row = new_steps // t if t > 0 else -1
        
        # Normal movement
        for dr, dc in directions:
            nr, nc = r + dr, c + dc
            
            if 0 <= nr < n and 0 <= nc < m:
                # Check if this row is blocked by dynamic wall
                if nr == blocked_row:
                    continue
                
                if (nr, nc, used_portal, new_steps) not in visited and grid[nr][nc] != '#':
                    visited.add((nr, nc, used_portal, new_steps))
                    q.append((nr, nc, used_portal, new_steps))
        
        # Portal teleportation
        if grid[r][c] == 'P' and not used_portal:
            for pr, pc in portals:
                if (pr, pc) != (r, c):
                    # Check if portal destination is blocked
                    if pr == blocked_row:
                        continue
                    
                    if (pr, pc, True, new_steps) not in visited:
                        visited.add((pr, pc, True, new_steps))
                        q.append((pr, pc, True, new_steps))
    
    return -1


if __name__ == "__main__":
    # Example from problem
    grid1 = [
        "S.#.",
        ".P..",
        "..#E",
        "P..."
    ]
    t = 2
    result1 = bfs_dynamic_walls(grid1, t)
    print(f"Part3 Example: {result1} steps (with dynamic walls, t={t})")
    
    # Test 2: Path gets blocked
    grid2 = [
        "S..",
        "..E"
    ]
    result2 = bfs_dynamic_walls(grid2, 3)  # t=3, row 1 blocked at step 3
    print(f"Part3 Test2: {result2} steps (path blocked scenario)")
    
    print("Day12 Part3 dynamic_walls.py: tests passed")
